dynamiq/nodes/agents/react.py

```diff
@@ -1851,62 +1848,100 @@ def _execute_tools(self, tools_data: list[dict], config: RunnableConfig, **kwarg
         Returns:
             str: Combined observation string with all tool results
         """
-        all_results = []
-
-        for tool_data in tools_data:
-            try:
-                tool_name = tool_data["name"]
-                tool_input = tool_data["input"]
-
-                tool = self.tool_by_names.get(self.sanitize_tool_name(tool_name))
-                if not tool:
-                    error_message = f"Unknown tool: {tool_name}. Please use only available tools."
-                    all_results.append({"tool_name": tool_name, "success": False, "result": error_message})
-                    continue
+        all_results: list[dict[str, Any]] = []
+
+        if not tools_data:
+            return ""
+
+        def _run_single_tool(tool_name: str, tool_input: dict) -> dict[str, Any]:
```

Comment:
Please move these inner functions to Agent level. Currently, it's too hard to read and understand flow